indispensable-cosef/src/commonMain/kotlin/at/asitplus/signum/indispensable/cosef/CoseSignedSerializer.kt

```diff
@@ -12,33 +14,45 @@ import kotlinx.serialization.encoding.Encoder
 
-class CoseSignedSerializer<P : Any?> : KSerializer<CoseSigned<P>> {
+class CoseSignedSerializer<P : Any?>(
+    private val parameterSerializer: KSerializer<P>,
+) : KSerializer<CoseSigned<P>> {
 
     @OptIn(InternalSerializationApi::class)
     override val descriptor: SerialDescriptor = buildSerialDescriptor("CoseSigned", StructureKind.LIST) {
         element("protectedHeader", ByteStringWrapperSerializer(CoseHeader.serializer()).descriptor)
         element("unprotectedHeader", CoseHeader.serializer().descriptor)
-        element("payload", ByteArraySerializer().descriptor)
+        element("payload", ByteStringWrapperSerializer(parameterSerializer).descriptor)
         element("signature", ByteArraySerializer().descriptor)
     }
 
-    override fun deserialize(decoder: Decoder): CoseSigned<P> {
-        return decoder.decodeStructure(descriptor) {
-            val protectedHeader = decodeSerializableElement(descriptor, 0, ByteStringWrapperSerializer(CoseHeader.serializer()))
-            val unprotectedHeader = decodeNullableSerializableElement(descriptor, 1, CoseHeader.serializer())
-            val payload = decodeNullableSerializableElement(descriptor, 2, ByteArraySerializer())
-            val signature = decodeSerializableElement(descriptor, 3, ByteArraySerializer())
-            CoseSigned(protectedHeader, unprotectedHeader, payload, signature)
+    override fun deserialize(decoder: Decoder): CoseSigned<P> = decoder.decodeStructure(descriptor) {
+        val protectedHeader =
+            decodeSerializableElement(descriptor, 0, ByteStringWrapperSerializer(CoseHeader.serializer()))
+        val unprotectedHeader = decodeNullableSerializableElement(descriptor, 1, CoseHeader.serializer())
+        val payload: ByteArray? = decodeNullableSerializableElement(descriptor, 2, ByteArraySerializer())
+        val signature: ByteArray = decodeSerializableElement(descriptor, 3, ByteArraySerializer())
+        runCatching {
+            val typedPayload: P? = payload?.let {
+                coseCompliantSerializer.decodeFromByteArray(parameterSerializer, it)
+            }
+            CoseSigned(protectedHeader, unprotectedHeader, typedPayload, signature)
+        }.getOrElse {
+            CoseSigned(protectedHeader, unprotectedHeader, payload as P, signature)
         }
     }
 
     override fun serialize(encoder: Encoder, value: CoseSigned<P>) {
         encoder.encodeStructure(descriptor) {
             encodeSerializableElement(descriptor, 0, ByteStringWrapperSerializer(CoseHeader.serializer()), value.protectedHeader)
             encodeNullableSerializableElement(descriptor, 1, CoseHeader.serializer(), value.unprotectedHeader)
-            encodeNullableSerializableElement(descriptor, 2, ByteArraySerializer(), value.payload)
+            if (value.payload != null && value.payload::class != ByteArray::class) {
+                // TODO Here, tag the result with 24 (decimal) = D818 (hex string)
```

Comment:
Maybe add a value class that does nothing but add the tag through an annotation and then wrap it here?

Reply:
I'm not sure what you mean exactly, but this doesn't work, because serializer for Any is not found:

```kotlin
@Serializable
internal data class TagWrapper(
    @ValueTags(24u)
    val value: ByteStringWrapper<*>
)
```